internal/server/websocketServer.go

```diff
@@ -0,0 +1,191 @@
+	mux.HandleFunc("/health", func(w http.ResponseWriter, r *http.Request) {
+		_, err := w.Write([]byte("OK"))
+		if err != nil {
+			return
+		}
+	})
+	return websocketServer
+}
+
+func (s *WebsocketServer) Run(ctx context.Context) error {
+	var wg sync.WaitGroup
+	var err error
+
+	websocketCtx, cancelWebsocket := context.WithCancel(ctx)
+	defer cancelWebsocket()
+
+	s.shardManager.RegisterWorker("wsServer", s.ioChan)
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		select {
+		case <-ctx.Done():
+		case <-s.shutdownChan:
+			err = diceerrors.ErrAborted
+			s.logger.Debug("Shutting down Websocket Server")
+		}
+
+		shutdownErr := s.websocketServer.Shutdown(websocketCtx)
+		if shutdownErr != nil {
+			s.logger.Error("Websocket Server shutdown failed:", slog.Any("error", err))
+			return
+		}
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		s.logger.Info("Websocket Server running", slog.String("port", s.websocketServer.Addr[1:]))
+		err = s.websocketServer.ListenAndServe()
+	}()
+
+	wg.Wait()
+	return err
+}
+
+func (s *WebsocketServer) WebsocketHandler(w http.ResponseWriter, r *http.Request) {
+	// upgrade http connection to websocket
+	conn, err := s.upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		return
+	}
+
+	// closing handshake
+	defer func() {
+		_ = conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseNormalClosure, "close 1000 (normal)"))
+		conn.Close()
+	}()
+
+	for {
+		// read incoming message
+		_, msg, err := conn.ReadMessage()
+		if err != nil {
+			sendTextMessage(conn, []byte("error: command reading failed"))
+			continue
+		}
+
+		// parse message to dice command
+		redisCmd, err := utils.ParseWebsocketMessage(msg)
+		if err == diceerrors.ErrEmptyCommand {
+			continue
+		} else if err != nil {
+			sendTextMessage(conn, []byte("error: parsing failed"))
+			continue
+		}
+
+		if redisCmd.Cmd == Abort {
+			close(s.shutdownChan)
+			break
+		}
+
+		if unimplementedCommandsWebsocket[redisCmd.Cmd] {
+			sendTextMessage(conn, []byte("Command is not implemented with Websocket"))
+			continue
+		}
+
+		// send request to Shard Manager
+		s.shardManager.GetShard(0).ReqChan <- &ops.StoreOp{
+			Cmd:         redisCmd,
+			WorkerID:    "wsServer",
+			ShardID:     0,
+			WebsocketOp: true,
+		}
+
+		// Wait for response
+		resp := <-s.ioChan
+		var rp *clientio.RESPParser
+		if resp.EvalResponse.Error != nil {
+			rp = clientio.NewRESPParser(bytes.NewBuffer([]byte(resp.EvalResponse.Error.Error())))
+		} else {
+			rp = clientio.NewRESPParser(bytes.NewBuffer(resp.EvalResponse.Result.([]byte)))
+		}
+
+		val, err := rp.DecodeOne()
+		if err != nil {
+			sendTextMessage(conn, []byte("error: decoding response"))
+			continue
+		}
+
+		respBytes, err := json.Marshal(val)
+		if err != nil {
+			sendTextMessage(conn, []byte("error: marshaling json response"))
+			continue
+		}
+
+		// Write response
+		sendTextMessage(conn, respBytes)
+	}
+}
+
+func sendTextMessage(conn *websocket.Conn, text []byte) {
```

Comment:
Shall we rename this to `func (conn *websocket.Conn) writeResponse(text []byte)`?

Reply:
Done.